### PR TITLE
Sprite Refactor

### PR DESCRIPTION
**Pull request by duncanspumpkin**
This modifies the size of rct_sprite to `0x200`. I've also moved a good number of functions to use the SpriteBase base class as then we can avoid a bunch of pointless casting.

I've also renamed `rct_vehicle` to `Vehicle` and renamed all of the other sprites that didn't have normal names.

In general I noticed a good few little mistakes along the way. I've tried to break up the PR into a bunch of smaller commits but I suspect it will be hard to follow no matter what.

### CODE REVIEW

File: src/openrct2/world/Sprite.h [hunk 1]
```diff
@@ -138,7 +138,7 @@ union rct_sprite
     rct_money_effect* AsMoneyEffect();
     Peep* AsPeep();
 };
-assert_struct_size(rct_sprite, 0x100);
+assert_struct_size(rct_sprite, 0x200);
```
Comment (tupaschoal) [inline src/openrct2/world/Sprite.h]: Out of curiosity, I thought that `structs` with these asserts couldn't be changed due to export/import limitations, isn't that true?

File: src/openrct2/world/Sprite.cpp [hunk 2]
```diff
@@ -689,26 +689,26 @@ void sprite_set_coordinates(int16_t x, int16_t y, int16_t z, rct_sprite* sprite)
  *
  *  rct2: 0x0069EDB6
  */
-void sprite_remove(rct_sprite* sprite)
+void sprite_remove(SpriteBase* sprite)
 {
-    auto peep = sprite->AsPeep();
+    auto peep = ((rct_sprite*)sprite)->AsPeep();
```
Comment (tupaschoal) [inline src/openrct2/world/Sprite.cpp]: This one sounds weird

File: src/openrct2/interface/InteractiveConsole.cpp [hunk 3]
```diff
@@ -1518,19 +1518,17 @@ static int32_t cc_mp_desync(InteractiveConsole& console, const arguments_t& argv
...
     for (int i = 0; i < MAX_SPRITES; i++)
     {
         rct_sprite* sprite = get_sprite(i);
         if (sprite->generic.sprite_identifier == SPRITE_IDENTIFIER_NULL)
             continue;
 
-        if (sprite->generic.sprite_identifier == SPRITE_IDENTIFIER_PEEP)
-            peeps.push_back(sprite);
-        else if (sprite->generic.sprite_identifier == SPRITE_IDENTIFIER_VEHICLE)
-            vehicles.push_back(sprite);
```
Comment (tupaschoal) [inline src/openrct2/interface/InteractiveConsole.cpp]: Why is the `vehicles` vector gone?

File: src/openrct2/world/Sprite.h (hunk 1 above)
Comment (duncanspumpkin) [reply src/openrct2/world/Sprite.h]: Not quite true. These are the in memory structs. We can change their sizes and structure but any new fields can't be saved as that would break export/import.

File: src/openrct2/world/Sprite.cpp (hunk 2 above)
Comment (duncanspumpkin) [reply src/openrct2/world/Sprite.cpp]: Yeah I really want to refactor so that `rct_sprite` is basically never used except when you need to know the max size of a sprite. For now this is just a stupid way of getting to the function I want.

File: src/openrct2/interface/InteractiveConsole.cpp (hunk 3 above)
Comment (duncanspumpkin) [reply src/openrct2/interface/InteractiveConsole.cpp]: It wasn't used for anything so thought there was no harm removing it. This function went through a few different versions during testing so there was a bit of dead code.

**Comment by IntelOrca** [comment]
What do you intend to do with the extra 0x100 bytes?

**Comment by duncanspumpkin** [comment]
> What do you intend to do with the extra 0x100 bytes?

Space for moving to bigger structures for the coordinates for one. I've chosen 0x100 just as a number that was bigger no real thought in it.

**Comment by IntelOrca** [comment]
@duncanspumpkin well I think it makes sense to keep it a power of 2

File: src/openrct2/GameStateSnapshots.cpp [hunk 4]
```diff
@@ -308,83 +308,82 @@ struct GameStateSnapshots : public IGameStateSnapshots
...
+        COMPARE_FIELD(Vehicle, var_44);
+        COMPARE_FIELD(Vehicle, mass);
+        COMPARE_FIELD(Vehicle, update_flags);
+        COMPARE_FIELD(Vehicle, swing_sprite);
+        COMPARE_FIELD(Vehicle, current_station);
+        COMPARE_FIELD(Vehicle, swinging_car_var_0);
+        COMPARE_FIELD(Vehicle, var_4E);
+        COMPARE_FIELD(Vehicle, status);
+        COMPARE_FIELD(Vehicle, sub_state);
         for (int i = 0; i < 32; i++)
```
Comment (Gymnasiast) [inline src/openrct2/GameStateSnapshots.cpp]: While we're at it, could you get rid of these hard coded maximums?

File: src/openrct2/world/Sprite.h (hunk 1 above)
Comment (tupaschoal) [reply src/openrct2/world/Sprite.h]: Ah, alright, thanks for explaining :)

File: src/openrct2/GameStateSnapshots.cpp (hunk 4 above)
Comment (duncanspumpkin) [reply src/openrct2/GameStateSnapshots.cpp]: I'll address that in a follow up pr.
